runtime/doc/lua.txt

```diff
@@ -2230,6 +2230,11 @@ vim.tbl_count({t})                                           *vim.tbl_count()*
 vim.tbl_deep_extend({behavior}, {...})                 *vim.tbl_deep_extend()*
     Merges recursively two or more tables.
 
+    Only empty tables or tables that are not |lua-list|s (indexed by
+    consecutive integers starting from 1) are merged. This is useful for
```

Comment:
For reference, the reverted implementation (the one present in 0.10.1) does not do this on "first level" of merging. I.e.:
 - `vim.tbl_deep_extend('force', {'a', 'b'}, {'c'})` -> `{ 'c', 'b' }`
 - `vim.tbl_deep_extend('force', {sub = {'a', 'b'} }, { sub = { 'c' } })` -> `{ sub = { 'c' } }`.

This was the whole point of #23654. Not sure if this still requires a fix or should be documented as intended.

Reply:
This is very likely intended/expected (note that tables of tables -- like plugin specs -- are also list-like). I'll try to document this.

Reply:
(And, yes, this is an inconsistency which would be better to fix; alas, we cannot do so without too much breakage, as it turns out. Hence we need a proper deprecation cycle, with a new API that is rational.)

Reply:
fyi, not sure if you're talking about lazy's plugin specs, but there @echasnovski's first example would yield `{'c'}`, which does indeed make more sense. So I'm also a proponent to have this fixed (either now, or in a later iteration)

Reply:
I was talking more generally, as an excuse not to get my hands too dirty. If you can promise to keep the general pitchforks at bay, I can look at extending `can_merge` to the top-level scope. That will likely take a bit more time since #dayjob will reassert itself powerfully soon.

Reply:
The most important thing right now is the PR in its current state.
In the context of option/config merging, the top level tables would not be list-like, so this problem does not happen there at least.

I'm willing to work on a `vim.merge` PR that includes this fix this week (after this PR gets merged).

Reply:
If either of you need / want testing or feedback from an end user perspective as this progresses, please don't hesitate to ask!